src/mcmc/Inference.jl

```diff
@@ -79,29 +79,61 @@ abstract type StaticHamiltonian <: Hamiltonian end
 abstract type AdaptiveHamiltonian <: Hamiltonian end
 
 """
-    ExternalSampler{S<:AbstractSampler}
+    ExternalSampler{S<:AbstractSampler,Unconstrained}
+
+Represents a sampler that is not an implementation of `InferenceAlgorithm`.
+
+The `Unconstrained` type-parameter is to indicate whether the sampler requires unconstrained space.
 
 # Fields
 $(TYPEDFIELDS)
 """
-struct ExternalSampler{S<:AbstractSampler} <: InferenceAlgorithm
+struct ExternalSampler{S<:AbstractSampler,AD<:ADTypes.AbstractADType,Unconstrained} <: InferenceAlgorithm
     "the sampler to wrap"
     sampler::S
+    "the automatic differentiation (AD) backend to use"
+    adtype::AD
+
+    @doc """
+        ExternalSampler(sampler::AbstractSampler, adtype::ADTypes.AbstractADType, unconstrained::Bool=true)
+
+    Wrap a sampler so it can be used as an inference algorithm.
+    """
+    function ExternalSampler(sampler::S, adtype::AD, unconstrained::Bool=true) where {S<:AbstractSampler,AD<:ADTypes.AbstractADType}
+        return new{S,AD,unconstrained}(sampler, adtype)
+    end
```

Comment:
Generally, the return type can't be inferred here. Maybe make unconstrained the first type parameter and only support
```julia
ExternalSampler{unconstrained}(sampler, adtype) where unconstrained = ExternalSampler{unconstrained,typeof(sampler),typeof(adtype)}(sampler, adtype)
ExternalSampler(sampler, adtype) = ExternalSampler{true}(sampler, adtype)
```

Reply:
Am aware, and am happy to make this change :+1: Buuuut the user will be using `externalsampler` to construct it anyways, which will never be inferrable either :confused: 

Buuut I agree with you;; we should at least have the _option_ of using a type-inferrable constructor so will make the change :+1:

Reply:
Maybe use `Val(true)`/`Val(false)` in `externalsampler` to make it fully inferrable?

Reply:
> Maybe use Val(true)/Val(false) in externalsampler to make it fully inferrable?

Don't like this as it requires users to be familiar with what `Val` is, which doesn't seem worth it given that `externalsampler` will never be used in "hot" code (it's really only going to be used before `sample(...)`, which is not type-stable anyways) :confused:

Reply:
Made the changes you wanted to the constructor (though leeft `externalsampler` as is)

Reply:
Making use of `Val` for constructor `ExternalSampler`, but leaving `externalsampler` as is I think